In [1]:
import numpy as np
import pandas as pd
import pickle

from sklearn.model_selection import train_test_split
from sklearn.model_selection import GridSearchCV
from sklearn.preprocessing import MinMaxScaler
from sklearn.utils import resample
from sklearn.metrics import classification_report,accuracy_score

from sklearn.svm import SVC
from sklearn.multioutput import MultiOutputClassifier

from sklearn.linear_model import SGDClassifier
from sklearn.neighbors import KNeighborsClassifier

In [2]:
    with open('features.pkl', 'rb') as file:
        X = pickle.load(file)
        X=np.array(X)

    with open('labels.pkl', 'rb') as file:
        Y = pickle.load(file)
        Y=np.array(Y)

In [33]:
columns_y=['y0','y1','y2','y3','y4','y5','y6','y7','y8','y9']
y_df=pd.DataFrame(Y,columns=columns_y)
    
columns_x=['age','gender','income','became_member_on']
x_df=pd.DataFrame(X,columns=columns_x)
    
for col in columns_y:
    
    print(col)
    table=pd.concat([x_df,y_df[col]],axis=1)
    df_majority = table[table[col]==0]
    df_minority = table[table[col]==1]
    
    # Upsample minority class
    df_minority_upsampled = resample(df_minority, 
                                     replace=True,   # sample with replacement
                                     n_samples=df_majority.shape[0]  # to match majority class
                                ) 
 
    # Combine majority class with upsampled minority class
    table = pd.concat([df_majority, df_minority_upsampled])
    
    scaler=MinMaxScaler()
    rescaled_data=scaler.fit_transform(table[columns_x])
    X_train, X_test, y_train, y_test = train_test_split(rescaled_data,\
                                                        table[col], test_size=0.2)

    y_test=np.array(y_test)
    y_train=np.array(y_train)
    
    #table[col].value_counts()
    
    #clf = SVC(kernel='rbf',class_weight='balanced',C=1, gamma='auto')

    clf = KNeighborsClassifier(n_neighbors=1)

    clf.fit(X_train, y_train)
    
    #compute y_predicted
    y_pred = clf.predict(X_test)
    #compute metrics
    print(classification_report(y_test,y_pred))

y0
              precision    recall  f1-score   support

           0       0.98      0.83      0.90      2477
           1       0.85      0.99      0.92      2549

    accuracy                           0.91      5026
   macro avg       0.92      0.91      0.91      5026
weighted avg       0.92      0.91      0.91      5026

y1
              precision    recall  f1-score   support

           0       0.99      0.82      0.90      2484
           1       0.85      0.99      0.91      2515

    accuracy                           0.91      4999
   macro avg       0.92      0.91      0.91      4999
weighted avg       0.92      0.91      0.91      4999

y2
              precision    recall  f1-score   support

           0       1.00      0.86      0.92      2643
           1       0.87      1.00      0.93      2560

    accuracy                           0.93      5203
   macro avg       0.93      0.93      0.93      5203
weighted avg       0.94      0.93      0.93      5203

y3
       

kernel=rbf c=1 gamma=auto

y0
              precision    recall  f1-score   support

           0       0.61      0.48      0.54      2546
           1       0.56      0.69      0.62      2480

    accuracy                           0.58      5026
   macro avg       0.59      0.58      0.58      5026
weighted avg       0.59      0.58      0.58      5026

y1
          precision    recall  f1-score   support

           0       0.67      0.51      0.58      2487
           1       0.61      0.75      0.67      2512

    accuracy                           0.63      4999
   macro avg       0.64      0.63      0.62      4999
weighted avg       0.64      0.63      0.62      4999

y2
              precision    recall  f1-score   support

           0       0.58      0.51      0.54      2574
           1       0.57      0.64      0.60      2629

    accuracy                           0.58      5203
   macro avg       0.58      0.57      0.57      5203
weighted avg       0.58      0.58      0.57      5203

y3
            precision    recall  f1-score   support

           0       0.61      0.37      0.46      2355
           1       0.55      0.77      0.64      2411

    accuracy                           0.57      4766
   macro avg       0.58      0.57      0.55      4766
weighted avg       0.58      0.57      0.55      4766

y4
             precision    recall  f1-score   support

           0       0.63      0.53      0.57      2711
           1       0.60      0.69      0.64      2736

    accuracy                           0.61      5447
   macro avg       0.61      0.61      0.61      5447
weighted avg       0.61      0.61      0.61      5447

y5
             precision    recall  f1-score   support

           0       0.60      0.46      0.52      2269
           1       0.56      0.70      0.62      2256

    accuracy                           0.58      4525
   macro avg       0.58      0.58      0.57      4525
weighted avg       0.58      0.58      0.57      4525

y6
             precision    recall  f1-score   support

           0       0.56      0.54      0.55      2151
           1       0.58      0.61      0.60      2307

    accuracy                           0.57      4458
   macro avg       0.57      0.57      0.57      4458
weighted avg       0.57      0.57      0.57      4458

y7
              precision    recall  f1-score   support

           0       0.60      0.49      0.54      2528
           1       0.59      0.69      0.63      2672

    accuracy                           0.59      5200
   macro avg       0.59      0.59      0.59      5200
weighted avg       0.59      0.59      0.59      5200

y8
             precision    recall  f1-score   support

           0       0.61      0.45      0.52      2622
           1       0.56      0.72      0.63      2609

    accuracy                           0.58      5231
   macro avg       0.59      0.58      0.58      5231
weighted avg       0.59      0.58      0.58      5231

y9
            precision    recall  f1-score   support

           0       0.58      0.51      0.54      2372
           1       0.57      0.64      0.60      2401

    accuracy                           0.57      4773
   macro avg       0.58      0.57      0.57      4773
weighted avg       0.58      0.57      0.57      4773

In [34]:
y_pred

array([1, 0, 1, ..., 1, 1, 1])

In [35]:
y_test

array([1, 0, 1, ..., 1, 1, 1])